ProjFS.Mac/PrjFSKextTests/ShouldHandleFileOpTests.mm

```diff
@@ -0,0 +1,298 @@
+}
+
+- (void)testProviderOffline {
+    PrjFSProviderUserClient userClient;
+    VirtualizationRootHandle testRootHandle = InsertVirtualizationRoot_Locked(
+        &userClient,
+        0,
+        self->repoRootVnode.get(),
+        self->repoRootVnode->GetVid(),
+        FsidInode{ self->repoRootVnode->GetMountPoint()->GetFsid(), self->repoRootVnode->GetInode() },
+        self->repoPath.c_str());
+    XCTAssertTrue(VirtualizationRoot_IsValidRootHandle(testRootHandle));
+
+    int pid;
+    // Fail when the provider is not online
+    s_virtualizationRoots[0].providerUserClient = nullptr;
+    XCTAssertFalse(
+        ShouldHandleFileOpEvent(
+            &self->perfTracer,
+            self->context,
+            self->testVnodeFile.get(),
+            nullptr, // path
+            KAUTH_FILEOP_RENAME,
+            true, // isDirectory,
+            &testRootHandle,
+            &pid));
+}
+
+- (void)testProviderInitiatedIO {
+    PrjFSProviderUserClient userClient;
+    VirtualizationRootHandle testRootHandle = InsertVirtualizationRoot_Locked(
+        &userClient,
+        0,
+        self->repoRootVnode.get(),
+        self->repoRootVnode->GetVid(),
+        FsidInode{ self->repoRootVnode->GetMountPoint()->GetFsid(), self->repoRootVnode->GetInode() },
+        self->repoPath.c_str());
+    XCTAssertTrue(VirtualizationRoot_IsValidRootHandle(testRootHandle));
+
+    // Fail when pid matches provider pid
+    MockProcess_Reset();
+    MockProcess_AddContext(self->context, 0 /*pid*/);
+    MockProcess_SetSelfPid(0);
+    MockProcess_AddProcess(0 /*pid*/, 1 /*credentialId*/, 1 /*ppid*/, "test" /*name*/);
+    int pid;
+    XCTAssertFalse(
+        ShouldHandleFileOpEvent(
+            &self->perfTracer,
+            self->context,
+            self->testVnodeFile.get(),
+            nullptr, // path
+            KAUTH_FILEOP_RENAME,
+            true, // isDirectory,
+            &testRootHandle,
+            &pid));
+}
+
+- (void)testVnodeCacheUpdated {
+    shared_ptr<vnode> testVnodeDirectory = self->testMount->CreateVnodeTree(self->repoPath + "/directory", VDIR);
+
+    VirtualizationRootHandle rootHandle;
+    int pid;
+    PrjFSProviderUserClient userClient;
+    VirtualizationRootHandle testRootHandle = InsertVirtualizationRoot_Locked(
+        &userClient,
+        0,
+        self->repoRootVnode.get(),
+        self->repoRootVnode->GetVid(),
+        FsidInode{ self->repoRootVnode->GetMountPoint()->GetFsid(), self->repoRootVnode->GetInode() },
+        self->repoPath.c_str());
+    XCTAssertTrue(VirtualizationRoot_IsValidRootHandle(testRootHandle));
+
+    // KAUTH_FILEOP_OPEN
+    XCTAssertTrue(
+        ShouldHandleFileOpEvent(
+            &self->perfTracer,
+            self->context,
+            self->testVnodeFile.get(),
+            nullptr, // path
+            KAUTH_FILEOP_OPEN,
+            false, // isDirectory,
+            &rootHandle,
+            &pid));
+    XCTAssertTrue(rootHandle == testRootHandle);
```

Comment:
While we're poking around in here, would it be worth changing these assertions to `XCTAssertEqual`?